ide/che-core-ide-jseditor/src/main/java/org/eclipse/che/ide/jseditor/client/debug/BreakpointManagerImpl.java

```diff
@@ -8,143 +8,235 @@
-                    debugger.deleteBreakpoint(activeFile, lineNumber, new AsyncCallback<Void>() {
-                        @Override
-                        public void onSuccess(final Void result) {
-                            breakpointsForPath.remove(breakpoint);
-                            breakpointRenderer.removeBreakpointMark(lineNumber);
-                        }
-
-                        @Override
-                        public void onFailure(final Throwable exception) {
-                            if (exception instanceof ServerException) {
-                                final ServerException e = (ServerException)exception;
-                                if (e.isErrorMessageProvided()) {
-                                    BreakpointManagerImpl.this.console.print(e.getMessage());
-                                    return;
-                                }
-                            }
-                            BreakpointManagerImpl.this.console.print("Can't delete breakpoint at line " + (lineNumber + 1));
-                        }
-                    });
-                    LOG.fine("Breakpoint removed.");
-                    if (breakpointsForPath.isEmpty()) {
-                        breakpoints.remove(activeFile.getPath());
-                    }
+                    // breakpoint already exists at given line
+                    deleteBreakpoint(activeFile, debugger, breakpoint);
                     return;
                 }
             }
         }
 
+        if (isCodeExecutable(activeFile, lineNumber)) {
+            addBreakpoint(activeFile, debugger, lineNumber);
+        }
+    }
+
+    /**
+     * Deletes breakpoint from the list and JVM.
+     * Removes breakpoint mark.
+     */
+    private void deleteBreakpoint(final VirtualFile activeFile,
+                                  final Debugger debugger,
+                                  final Breakpoint breakpoint) {
+        doDeleteBreakpoint(activeFile, breakpoint);
+
+        debugger.deleteBreakpoint(activeFile, breakpoint.getLineNumber(), new AsyncCallback<Void>() {
+            @Override
+            public void onSuccess(final Void result) {
+            }
+
+            @Override
+            public void onFailure(final Throwable exception) {
+                if (debuggerState == CONNECTED) {
+                    doAddBreakpoint(activeFile, breakpoint);
+                }
+            }
+        });
+    }
+
+    /**
+     * Deletes breakpoint from the list.
+     */
+    private void doDeleteBreakpoint(VirtualFile activeFile, Breakpoint breakpoint) {
+        BreakpointRenderer breakpointRenderer = getBreakpointRendererForFile(activeFile);
+        if (breakpointRenderer != null) {
+            breakpointRenderer.removeBreakpointMark(breakpoint.getLineNumber());
+        }
+
+        String path = breakpoint.getPath();
+
+        List<Breakpoint> fileBreakpoints = breakpoints.get(path);
```

Comment:
This is a good point. I consider it like a different code style. In general case I prefer to keep a code line as simple as possible. It simplifies understanding the entire code as well the root of exceptions.